### PR TITLE
HPCC-27522 Improve logging of initial LDAP Connect

### PR DESCRIPTION
**Pull request by RussWhitehead**
1) Don't ever try unsecure connect (389) if secure was config choice (636)
2) Don't log ananymous bind errors, just quietly try with creds
3) Added a couple logfile message improvements

Signed-off-by: Russ Whitehead <russwhitehead@yahoo.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [x] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by RussWhitehead** [comment]
@wangkx Please review

File: system/security/LdapSecurity/ldaputils.cpp [hunk 1]
```diff
@@ -231,9 +231,6 @@ LDAP* LdapUtils::ldapInitAndSimpleBind(const char* ldapserver, const char* userD
     *err = LdapSimpleBind(ld, timeout, (char*)userDN, (char*)pwd);
     if (*err != LDAP_SUCCESS)
     {
-        DBGLOG("LdapSimpleBind error (%d) - '%s' for admin user %s", *err, ldap_err2string(*err), isEmptyString(userDN) ? "NULL" : userDN);
-        if (!isEmptyString(userDN))
-            DBGLOG("Please make sure your LDAP configuration 'systemBasedn' contains the complete path, including the complete 'dc=domainComponent'");
```
Comment (wangkx) [inline system/security/LdapSecurity/ldaputils.cpp:236]: If not logged here, is it logged (or exception thrown) on the caller side?

File: system/security/LdapSecurity/ldaputils.cpp [hunk 2]
```diff
@@ -251,21 +248,16 @@ int LdapUtils::getServerInfo(const char* ldapserver, const char* userDN, const c
...
-
-    if(nullptr == ld)
-    {
-        DBGLOG("ldap bind error (%d) - %s", err, ldap_err2string(err));
-
-        // for new versions of openldap, version 2.2.*
-        if(err == LDAP_PROTOCOL_ERROR)
-            DBGLOG("If you're trying to connect to an OpenLdap server, make sure you have \"allow bind_v2\" enabled in slapd.conf");
+            // for new versions of openldap, version 2.2.*
+            if(err == LDAP_PROTOCOL_ERROR  &&  stype != ACTIVE_DIRECTORY)
```
Comment (wangkx) [inline system/security/LdapSecurity/ldaputils.cpp:256]: I guess that stype != ACTIVE_DIRECTORY only for openldap.

**Review by wangkx — COMMENTED** [review]
@RussWhitehead a few questions/comments.

File: system/security/LdapSecurity/ldaputils.cpp (hunk 1 above)
Comment (RussWhitehead) [reply system/security/LdapSecurity/ldaputils.cpp:236]: This behavior is unchanged. This routine logs the error, and the caller (CLdapConfig ctor) throws an exception (see ldapconnection.cpp, line 479)

        if(rc != LDAP_SUCCESS)
        {
            throw MakeStringException(-1, "getServerInfo error - %s", ldap_err2string(rc));
        }

**Comment by RussWhitehead** [comment]
@wangkx Please rereview

File: system/security/LdapSecurity/ldaputils.cpp (hunk 1 above)
Comment (wangkx) [reply system/security/LdapSecurity/ldaputils.cpp:236]: > This routine logs the error, .

I may misunderstand. If you remove those 3 lines, where do you log the error?

Comment (RussWhitehead) [reply system/security/LdapSecurity/ldaputils.cpp:236]: I no longer log the error from the anonymous bind  (see PR comment #2)
     "Don't log ananymous bind errors, just quietly try with creds"
If the anonymous bind fails, I quietly retry with the provided creds. If that one fails , I log it on line 253
     DBGLOG("ldap bind error (%d) - %s", err, ldap_err2string(err));
It's up to the caller to throw or return the error, but in this case the caller throws

**Review by wangkx — APPROVED** [review]
@RussWhitehead looks fine.

**Comment by RussWhitehead** [comment]
@richardkchapman Please merge

**Pull request merged**